SetupDataPkg/Library/ConfigKnobShimLib/ConfigKnobShimLibCommon.c

```diff
@@ -0,0 +1,89 @@
+/** @file
+  Common functionality for the library interface for an OEM config policy creator to call into to fetch overrides
+  for config values.
+
+  Copyright (c) Microsoft Corporation.
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+#include <Uefi.h>
+#include <Library/BaseLib.h>
+#include <Library/DebugLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/ConfigKnobShimLib.h>
+#include "ConfigKnobShimLibCommon.h"
+
+/**
+  GetConfigKnobOverride searches for an override to the given config knob.
+
+  If the config override is found and the data size matches ConfigKnobDataSize, ConfigKnobData will be written with
+  the override value.
+
+  If the config override is not found or another error occurs, ConfigKnobData will not be written with the override
+  value.
+
+  This function is only expected to be called from an OEM config policy creator.
+
+  @param[in]  ConfigKnobGuid      The GUID of the requested config knob.
+  @param[in]  ConfigKnobName      The name of the requested config knob.
+  @param[out] ConfigKnobData      The retrieved data of the requested config knob. The caller will allocate memory
+                                  for this buffer and the caller is responsible for freeing it.
+  @param[in] ConfigKnobDataSize   The allocated size of ConfigKnobData. This will be set to the correct value for the
+                                  size of ConfigKnobData in the success case.
+
+  @retval EFI_INVALID_PARAMETER   Input argument is null.
+  @retval EFI_BAD_BUFFER_SIZE     Override data size does not match config knob data size.
+  @retval EFI_NOT_FOUND           Override not found for this knob.
+  @retval !EFI_SUCCESS            Other error in finding an override to this config knob
+  @retval EFI_SUCCESS             The operation succeeds.
+
+**/
+EFI_STATUS
+EFIAPI
+GetConfigKnobOverride (
+  IN EFI_GUID  *ConfigKnobGuid,
+  IN CHAR16    *ConfigKnobName,
+  OUT VOID     *ConfigKnobData,
+  IN UINTN     ConfigKnobDataSize
+  )
+{
+  EFI_STATUS  Status;
+  UINTN       VariableSize = 0;
+
+  if ((ConfigKnobGuid == NULL) || (ConfigKnobName == NULL) || (ConfigKnobData == NULL) ||
+      (ConfigKnobDataSize == 0))
+  {
+    DEBUG ((DEBUG_ERROR, "%a: Invalid parameter!\n", __FUNCTION__));
+    Status = EFI_INVALID_PARAMETER;
+    goto Exit;
+  }
+
+  // Check size in variable storage
+  Status = GetConfigKnobFromVariable (ConfigKnobGuid, ConfigKnobName, NULL, &VariableSize);

```

Comment:
VariableSize is checked before status is checked?

Reply:
Yeah, the thought process there was to only have one print statement for not finding a variable. So, we check and see if the VariableSize is what we expect, if it isn't, no matter what the status is, we want to not use this variable. If it is the size we expect and the status is EFI_BUFFER_TOO_SMALL, we want to fetch the variable now that we've confirmed the size.

Then, if that fetch fails (or if the original size didn't match) we print at line 76.

All that being said, I'm fine for this to be rearranged, as long as the same guarantees are there.

Reply:
In unit testing, I saw a case where we would fetch the variable twice even when we know it isn't there. Fixed to your suggestion.